[FLINK-21258] Add Canceling state for DeclarativeScheduler

rmetzger opened the pull request:
This PR is based on https://github.com/apache/flink/pull/14879

## What is the purpose of the change

[Declarative Scheduler](https://cwiki.apache.org/confluence/display/FLINK/FLIP-160%3A+Declarative+Scheduler) consists of a number of internal states. 

Note that this change is currently not usable as-is, as the other parts of declarative scheduler are not merged yet (See for the prototype this PR is based on: https://github.com/tillrohrmann/flink/tree/declarative-scheduler) 


## Verifying this change

- The change is adding unit tests.
- Note that integration tests for the declarative scheduler will cover additional functionality.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: (**yes** / no / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

Will be handled in a separate PR.

tillrohrmann commented on flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/declarative/CancelingTest.java:
```diff
@@ -0,0 +1,171 @@
+    public void testTransitionToFinishedWhenCancellationCompletes() throws Exception {
+        try (MockStateWithExecutionGraphContext ctx = new MockStateWithExecutionGraphContext()) {
+            Canceling canceling = createCancelingState(ctx);
+
```
here we don't have to call `onEnter`?

tillrohrmann commented on flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/declarative/CancelingTest.java:
```diff
@@ -0,0 +1,171 @@
+    @Test
+    public void testTransitionToFinishedWhenCancellationCompletes() throws Exception {
+        try (MockStateWithExecutionGraphContext ctx = new MockStateWithExecutionGraphContext()) {
+            Canceling canceling = createCancelingState(ctx);
```
It is a bit hidden why this test works. I assume it is the case because the `ExecutionGraph` does not contain any `Executions` which are in `DEPLOYING` or `RUNNING` state. That's why a `ExecutionGraph.cancel` can directly cancel all `Executions`. Maybe we can make this a bit more explicit.

tillrohrmann commented on flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/declarative/CancelingTest.java:
```diff
@@ -0,0 +1,171 @@
+    public void testTransitionToSuspend() throws Exception {
+        try (MockStateWithExecutionGraphContext ctx = new MockStateWithExecutionGraphContext()) {
+            Canceling canceling = createCancelingState(ctx);
+
```
I assume that you don't call `onEnter` because it would make the state go to `Finished`. I think that we should stick to the convention of `State` and call `onEnter`. Otherwise we risk that we are testing something which is impossible to reach in production. The proper solution would be to ensure that the `ExecutionGraph` does not directly go to `CANCELLED` when `onEnter` is called.

tillrohrmann commented on flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/declarative/CancelingTest.java:
```diff
@@ -0,0 +1,171 @@
+                                    ExecutionState.FAILED,
+                                    new RuntimeException()));
+            canceling.updateTaskExecutionState(update);
+            ctx.assertNoStateTransition();
```
We should also assert that the `ExecutionGraph` has been updated. Otherwise a passing implementation of the `Canceling.updateTaskExecutionState` could be an empty method.

Technically, we would also have to do this for the other `StateWithExecutionGraph` sub classes.

tillrohrmann commented on flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/declarative/CancelingTest.java:
```diff
@@ -0,0 +1,171 @@
+        }
+
+        @Override
+        public void notifyGlobalFailure(Throwable t) {
+            canceling.handleGlobalFailure(t);
+        }
```
For which test do we need this implementation?

rmetzger replied:
For the `testTaskFailuresAreIgnored()` test.

I'm actually proposing to remove that test. It doesn't belong here. We need to test the proper handling of the `updateTaskExecutionState()` on the DeclarativeScheduler itself (which registers an InternalFailuresListener and forwards failures to the `handleGlobalFailure()` method of the State).
Since we are testing the `handleGlobalFailure()` in a unit test, having a test for `updateTaskExecutionState()` here doesn't make much sense.

@zentol: Can we add such a test to the DeclarativeScheduler skeleton PR?

tillrohrmann replied:
If I am not mistaken, then we don't need an `InternalFailuresListener` for testing the `updateTaskExecutionState`. I thought that his listener is only used when a failure in some other operation (e.g. `deploy` occurs). Testing that `updateTaskExecutionState` stays in the `CANCELING` state makes sense to me.

tillrohrmann reviewed (changes requested):
Thanks for creating this PR @rmetzger. The changes look already quite good. I had a couple of comments concerning the tests. Please take a look.

rmetzger commented:
I rebased this change to the latest master (thus fewer commits are included), and addressed all comments
Once you confirm that the `testTaskFailuresAreIgnored()` doesn't make sense, I'll remove it.

I once again introduced a MockExecutionGraph for this test. But extracting an interface from the ExecutionGraph is a really involved change that would not be worth it for my needs in this PR.
It is easier to introduce a ExecutionGraph interface in a separate change.

tillrohrmann reviewed (approved):
Thanks for updating this PR @rmetzger. I think it makes sense to keep `testTaskFailuresAreIgnored` because we want to ensure that we don't leave the state on a failed task state update. Moreover, I think we shouldn't need `TestInternalFailuresListener`.

rmetzger commented:
Thanks for your review!

I'll remove the `TestInternalFailuresListener` and mock the ExecutionGraph to intercept the `failGlobal()` method, then merge the change.
